### PR TITLE
templating tests and bugfixes

### PR DESCRIPTION
**Pull request by fkleuver**
# Pull Request

## 📖 Description

This PR is a first step towards a more maintenance-friendly approach to bulk testing input combinations.

There's also a small "bugfix" for an issue uncovered by the generated tests, that is the root element's children weren't being removed on the stop task. A pretty uncommon scenario that probably wouldn't have ever bothered anyone (except for me with integration testing), but easy enough to fix that I found it worth fixing. The `element.$au` property might also very well come in handy in other situations.

### 🎫 Issues

This is part of https://github.com/aurelia/aurelia/issues/89

## 👩‍💻 Reviewer Notes

The generated tests live in the `.generated.spec.ts` files (at the time of writing only `template-compiler.generated.spec.ts`)

## 📑 Test Plan

The goals of the generated test code are pretty straight-forward:
- make sure at least some tests fail on even minor code changes (generate near-redundant combinations of small variations)
- make failing tests easy to find (sequential numbering is in place)
- make failing tests easy to debug (no chained calls, no loops, one test per `it`)
- make failing tests easy to fix (change a small piece in the generator code to fix lots of tests)

## ⏭ Next Steps

At the moment the generator code in itself contains a lot of copy-paste work with small modifications. This needs to get cleaner, but I'd like to drive this effort in a pragmatic manner. It should quickly and easily produce a high test density in complex areas that have a high chance of misbehaving in more exotic setups.

I may remove some of the existing test loops when there is no question they are fully redundant.

The current tests are mostly static. An important (and somewhat tricky) next step will be to test bound data, and data that mutates.

Other good candidates for generated tests:
- Bindings
- Observers
- View, customElement, customAttribute lifecycle hooks
- Most things, really

The plan is not to replace *all* tests with generated tests, but rather the tedious ones that are hard to debug (e.g. the loops). They should represent the thorough variants of a handful of carefully hand-crafted "model" tests, so that it's always easy to branch out / add more tests without having to dig into the generator code.

### CODE REVIEW

**Comment by fkleuver** [comment]
@EisenbergEffect @bigopon this is still WIP, but please have a look at the generated tests and let me know what you think before more of the same stuff will land.

**Comment by bigopon** [comment]
@fkleuver This is freaking awesome. First glance at the test "source", things are way nicer. Very nice!

**Comment by EisenbergEffect** [comment]
Very nice!

File: packages/runtime/src/dom.ts [hunk 1]
```diff
@@ -6,6 +6,10 @@ export const TEXT_NODE = 3;
 export const COMMENT_NODE = 8;
 export const DOCUMENT_FRAGMENT_NODE = 11;
 
+function isRenderLocation(node: INode): node is IRenderLocation {
+  return node.textContent === 'au-end';
```
Comment (bigopon) [inline packages/runtime/src/dom.ts]: I've not followed the movement around this so this surprises me. Can you give some pointers @fkleuver

Comment (fkleuver) [reply packages/runtime/src/dom.ts]: I've renamed `au-loc` to `au-start` and `au-end` since there is now "start" counterpart to a RenderLocation. The original one (which always comes at the end) is `au-end`. `node.textContent` works the same on a comment like on any other node and will return that value.

File: packages/runtime/src/dom.ts [hunk 2]
```diff
@@ -310,47 +324,122 @@ export class FragmentNodeSequence implements INodeSequence {
...
   }
 
   public findTargets(): ArrayLike<INode> {
     // tslint:disable-next-line:no-any
-    return (<any>this.fragment).querySelectorAll('.au');
+    return this.targets;
   }
 
-  public insertBefore(refNode: INode): void {
+  public insertBefore(refNode: IRenderLocation): void {
```
[the remark this answers is not in the document]
Comment (fkleuver) [reply packages/runtime/src/dom.ts]: Well the `IRenderLocation` is just an `INode` with some additional optional properties, so from a typing perspective you can pass in an `INode` without it complaining. The public API (the interface) still uses `INode`. So I figured having it `IRenderLocation` on this implementation would save some explicit casting to access the properties

File: packages/runtime/src/dom.ts [hunk 3]
```diff
@@ -310,47 +324,122 @@ export class FragmentNodeSequence implements INodeSequence {
...
+  private targets: ArrayLike<INode>;
+
+  private start: IRenderLocation;
+  private end: IRenderLocation;
 
   constructor(fragment: IDocumentFragment) {
     this.fragment = fragment;
+    // querySelectorAll vs getElementsByClassName benchmark: http://jsben.ch/X4Gek
+    // tslint:disable-next-line:no-any
+    const targetNodeList = (<any>fragment).getElementsByClassName('au');
```
[the remark this answers is not in the document]
Comment (fkleuver) [reply packages/runtime/src/dom.ts]: Wow, I'm not sure how that slipped through. I was pretty sure the tests passed but now running them again locally they failed miserably. Well I reverted it again. I was wondering why you'd miss something like that, but of course you didn't :P

**Comment by fkleuver** [comment]
You're correct, it doesn't. Must have been a brain fart where I thought the tests passed with it, but turns out they didn't, so I reverted that again.

**Comment by fkleuver** [comment]
@EisenbergEffect @bigopon This is ready for review. I've done a number of important bugfixes that the tests helped me uncover and solve (I think this is a nice point for merging and I'd like to not pollute the scope too much, keep it reviewable):
- Root app element children were not removed on aurelia.stop, fixed this in the HostProjector: https://github.com/aurelia/aurelia/pull/245/commits/46f1d5f34c0701709508ffbde5c6162954ed264b
- As you already know, use start + end comment to correctly remove template controller elements in a deterministic manner: https://github.com/aurelia/aurelia/pull/245/commits/43005e701949c04f8babd71fe4e66c0e829539f5 and https://github.com/aurelia/aurelia/pull/245/commits/8fbe26b1d99fff72efe8747c16e30f122eec477d
- Use EmptyNodeSequence if the fragment has no children (e.g. empty `if`): https://github.com/aurelia/aurelia/pull/245/commits/d600da4dab48b7240b26ecfa8334626a88761f2b
- Only invoke the change handler on `with` when it is bound, fixes a bunch of tests related to with: https://github.com/aurelia/aurelia/pull/245/commits/935eedbf660af26fec64b774bb634608f13af188
- Set the `Text.textContent` to empty if no value is bound to it (prevent rendering unnecessary whitespace): https://github.com/aurelia/aurelia/pull/245/commits/783172607ad5b2be5095f7513e97c57a1705ec22
- Replace the host element for containerless with a marker element, making it behave in a similar fashion to template controllers: https://github.com/aurelia/aurelia/pull/245/commits/75bcaa5b68dd9fa67cd3cc7b25489751be629a87 (otherwise it didn't really work)
- Fix ShadowDOM (attachShadow doesn't work on custom elements apparently): https://github.com/aurelia/aurelia/pull/245/commits/18d2e1521f0f840d8196843136dd486d25915a7e

**Comment by bigopon** [comment]
> As you already know, use start + end comment to correctly remove template controller elements in a deterministic manner: 43005e7 and 8fbe26b

I always wanted to do it this way for vCurrent, as naked nested anchors are incapable of preserving their scopes. The work around this related to targets / fragment node sequence are also nice. This makes vNext feel more complete.

For the test, I think we will be able to confidently to play whack-a-mole with edge cases without worrying much.

Love it 😃

**Comment by fkleuver** [comment]
Changing this back to WIP because I found a pretty tricky bug in the way the lifecycles behave, the solution to which is actually my earlier proposal to integrate them all. So I suppose the time has come to integrate them.

**Comment by fkleuver** [comment]
Integrating the lifecycles is a pandora's box. I made an attempt and ended up with 2k replacements across 80 files within 5 minutes. I find the idea I came up with kind of appealing though, but it needs more work and thought. It's a huge undertaking and I'd rather try to do it in smaller steps, but I'm not sure how yet since it's all so interwoven.

In any case, I implemented a "hacky" workaround that solves the specific issue I ran into: https://github.com/aurelia/aurelia/pull/245/commits/6483ed9d03d11abe797fa53aed64ec942da33e58
Based on that I think it's better to continue with tests and other things first (and/or in parallel) before performing surgery on the lifecycles. We need much denser test coverage on it to make sure the refactoring won't cause other issues on the long run.

The generated mutation tests are a first step in that direction. Those have been quick to expose a couple serious issues already so they seem like a good investment of time at the moment. To get the core more robust. These are the kinds of tests I've been looking forward to from the very beginning.

If you're curious what the first rough draft of integrated lifecycles looked like - please do look past the various minor issues it may have. Looking at the broader picture - they're all aware of each other and are guaranteed to happen in a certain order, an attach will always make sure all changes and binds are done, etc:

(Promises could fit in there too)

```ts
export const Lifecycle = {
  flushing: <boolean>false,
  flushChangesDepth: 0,
  flushChangesHead: <ILifecycleFlushChanges>null,
  flushChangesTail: <ILifecycleFlushChanges>null,
  queueFlushChanges(requestor: ILifecycleFlushChanges, flags: LifecycleFlags): void {
    if (requestor.$nextFlushChanges) {
      return;
    }
    requestor.$flushChangesFlags = flags;
    requestor.$nextFlushChanges = marker;
    if (Lifecycle.flushChangesHead === null) {
      Lifecycle.flushChangesHead = requestor;
    } else {
      Lifecycle.flushChangesTail.$nextFlushChanges = requestor;
    }
    Lifecycle.flushChangesTail = requestor;
    ++Lifecycle.flushChangesDepth;
  },
  invokeFlushChangesCallbacks(): void {
    if (--Lifecycle.flushChangesDepth === 0) {
      if (Lifecycle.flushing) {
        return;
      }
      while (Lifecycle.flushChangesHead !== null) {
        let current = Lifecycle.flushChangesHead;
        let next: ILifecycleFlushChanges;
        Lifecycle.flushChangesHead = Lifecycle.flushChangesTail = null;
        while (current && current !== marker) {
          current.flushChanges(current.$flushChangesFlags);
          next = current.$nextFlushChanges;
          current.$nextFlushChanges = null;
          current = next;
        }
      }
      if (Lifecycle.boundDepth === 0) {
        Lifecycle.invokeBoundCallbacks();
      }
    }
  },

  boundDepth: 0,
  boundHead: <ILifecycleBound>null,
  boundTail: <ILifecycleBound>null,
  queueBound(requestor: ILifecycleBound, flags: LifecycleFlags): void {
    if (requestor.$nextBound) {
      return;
    }
    requestor.$boundFlags = flags;
    requestor.$nextBound = marker;
    if (Lifecycle.boundHead === null) {
      Lifecycle.boundHead = requestor;
    } else {
      Lifecycle.boundTail.$nextBound = requestor;
    }
    Lifecycle.boundTail = requestor;
    ++Lifecycle.boundDepth;
  },
  unqueueBound(): void {
    if (--Lifecycle.boundDepth === 0) {
      if (Lifecycle.flushChangesDepth > 0) {
        return;
      }
      Lifecycle.invokeBoundCallbacks();
    }
  },
  invokeBoundCallbacks(): void {
    let current = Lifecycle.boundHead;
    let next: ILifecycleBound;
    Lifecycle.boundHead = Lifecycle.boundTail = null;
    while (current && current !== marker) {
      current.bound(current.$boundFlags);
      next = current.$nextBound;
      current.$nextBound = null;
      current = next;
    }
    if (Lifecycle.attachedDepth === 0) {
      Lifecycle.invokeAttachedCallbacks();
    }
  },

  mountDepth: 0,
  mountHead: <ILifecycleMount>null,
  mountTail: <ILifecycleMount>null,
  queueMount(requestor: ILifecycleMount, flags: LifecycleFlags): void {
    if (requestor.$nextMount) {
      return;
    }
    requestor.$mountFlags = flags;
    requestor.$nextMount = marker;
    if (Lifecycle.mountHead === null) {
      Lifecycle.mountHead = requestor;
    } else {
      Lifecycle.mountTail.$nextMount = requestor;
    }
    Lifecycle.mountTail = requestor;
    ++Lifecycle.mountDepth;
  },
  unqueueMount(): void {
    if (--Lifecycle.mountDepth === 0) {
      if (Lifecycle.boundDepth > 0 || Lifecycle.flushChangesDepth > 0) {
        return;
      }
      if (Lifecycle.flushChangesDepth > 0) {
        Lifecycle.invokeFlushChangesCallbacks();
      }
      Lifecycle.invokeMountCallbacks();
    }
  },
  invokeMountCallbacks(): void {
    let current = Lifecycle.mountHead;
    let next: ILifecycleMount;
    Lifecycle.mountHead = Lifecycle.mountTail = null;
    while (current && current !== marker) {
      current.$mount(current.$mountFlags);
      next = current.$nextMount;
      current.$nextMount = null;
      current = next;
    }
    if (Lifecycle.attachedDepth === 0) {
      Lifecycle.invokeAttachedCallbacks();
    }
  },

  attachedDepth: 0,
  attachedHead: <ILifecycleAttached>null,
  attachedTail: <ILifecycleAttached>null,
  queueAttached(requestor: ILifecycleAttached, flags: LifecycleFlags): void {
    if (requestor.$nextAttached) {
      return;
    }
    requestor.$attachedFlags = flags;
    requestor.$nextAttached = marker;
    if (Lifecycle.attachedHead === null) {
      Lifecycle.attachedHead = requestor;
    } else {
      Lifecycle.attachedTail.$nextAttached = requestor;
    }
    Lifecycle.attachedTail = requestor;
    ++Lifecycle.attachedDepth;
  },
  unqueueAttached(): void {
    if (--Lifecycle.attachedDepth === 0) {
      if (Lifecycle.mountDepth > 0 || Lifecycle.boundDepth > 0 || Lifecycle.flushChangesDepth > 0) {
        return;
      }
      Lifecycle.invokeAttachedCallbacks();
    }
  },
  invokeAttachedCallbacks(): void {
    let current = Lifecycle.attachedHead;
    let next: ILifecycleAttached;
    Lifecycle.attachedHead = Lifecycle.attachedTail = null;
    while (current && current !== marker) {
      current.attached(current.$attachedFlags);
      next = current.$nextAttached;
      current.$nextAttached = null;
      current = next;
    }
  },

  unmountDepth: 0,
  unmountHead: <ILifecycleUnmount>null,
  unmountTail: <ILifecycleUnmount>null,
  queueUnmount(requestor: ILifecycleUnmount, flags: LifecycleFlags): void {
    if (requestor.$nextUnmount) {
      return;
    }
    requestor.$unmountFlags = flags;
    requestor.$nextUnmount = marker;
    if (Lifecycle.unmountHead === null) {
      Lifecycle.unmountHead = requestor;
    } else {
      Lifecycle.unmountTail.$nextUnmount = requestor;
    }
    Lifecycle.unmountTail = requestor;
    ++Lifecycle.unmountDepth;
  },
  unqueueUnmount(): void {
    if (--Lifecycle.unmountDepth === 0) {
      let current = Lifecycle.unmountHead;
      let next: ILifecycleUnmount;
      Lifecycle.unmountHead = Lifecycle.unmountTail = null;
      while (current && current !== marker) {
        current.$unmount(current.$unmountFlags);
        next = current.$nextUnmount;
        current.$nextUnmount = null;
        current = next;
      }
    }
  },

  detachedDepth: 0,
  detachedHead: <ILifecycleDetached>null,
  detachedTail: <ILifecycleDetached>null,
  queueDetached(requestor: ILifecycleDetached, flags: LifecycleFlags): void {
    if (requestor.$nextDetached) {
      return;
    }
    requestor.$detachedFlags = flags;
    requestor.$nextDetached = marker;
    if (Lifecycle.detachedHead === null) {
      Lifecycle.detachedHead = requestor;
    } else {
      Lifecycle.detachedTail.$nextDetached = requestor;
    }
    Lifecycle.detachedTail = requestor;
    ++Lifecycle.detachedDepth;
  },
  unqueueDetached(): void {
    if (--Lifecycle.detachedDepth === 0) {
      let current = Lifecycle.detachedHead;
      let next: ILifecycleDetached;
      Lifecycle.detachedHead = Lifecycle.detachedTail = null;
      while (current && current !== marker) {
        current.detached(current.$detachedFlags);
        next = current.$nextDetached;
        current.$nextDetached = null;
        current = next;
      }
    }
  },

  unboundDepth: 0,
  unboundHead: <ILifecycleUnbound>null,
  unboundTail: <ILifecycleUnbound>null,
  queueUnbound(requestor: ILifecycleUnbound, flags: LifecycleFlags): void {
    if (requestor.$nextUnbound) {
      return;
    }
    requestor.$unboundFlags = flags;
    requestor.$nextUnbound = marker;
    if (Lifecycle.unboundHead === null) {
      Lifecycle.unboundHead = requestor;
    } else {
      Lifecycle.unboundTail.$nextUnbound = requestor;
    }
    Lifecycle.unboundTail = requestor;
    ++Lifecycle.unboundDepth;
  },
  unqueueUnbound(): void {
    if (--Lifecycle.unboundDepth === 0) {
      let current = Lifecycle.unboundHead;
      let next: ILifecycleUnbound;
      Lifecycle.unboundHead = Lifecycle.unboundTail = null;
      while (current && current !== marker) {
        current.unbound(current.$unboundFlags);
        next = current.$nextUnbound;
        current.$nextUnbound = null;
        current = next;
      }
    }
  },
};
```

It would completely replace the lifecycle controllers and change set, and every lifecycle hook would take `LifecycleFlags`:

```ts
export enum LifecycleFlags {
  none                   = 0b0000_0000000000000_000_00,
  mustEvaluate           = 0b1000_0000000000000_000_00,
  mutation               = 0b0000_0000000000000_000_11,
  isCollectionMutation   = 0b0000_0000000000000_000_01,
  isInstanceMutation     = 0b0000_0000000000000_000_10,
  update                 = 0b0000_0000000000000_111_00,
  updateTargetObserver   = 0b0000_0000000000000_001_00,
  updateTargetInstance   = 0b0000_0000000000000_010_00,
  updateSourceExpression = 0b0000_0000000000000_100_00,
  from                   = 0b0000_1111111111111_000_00,
  fromFlushChanges       = 0b0000_0000000000001_000_00,
  fromStartTask          = 0b0000_0000000000010_000_00,
  fromStopTask           = 0b0000_0000000000100_000_00,
  fromBind               = 0b0000_0000000001000_000_00,
  fromUnbind             = 0b0000_0000000010000_000_00,
  fromDOMEvent           = 0b0000_0000000100000_000_00,
  fromObserverSetter     = 0b0000_0000001000000_000_00,
  fromBindableHandler    = 0b0000_0000010000000_000_00,
  fromAttach             = 0b0000_0000100000000_000_00,
  fromDetach             = 0b0000_0001000000000_000_00,
  fromMount              = 0b0000_0010000000000_000_00,
  fromUnmount            = 0b0000_0100000000000_000_00,
  fromRelease            = 0b0000_1000000000000_000_00,
}
```

As said, very rough, but that's the direction I had in mind.

So I'd like to make another attempt, for now, to wrap up this PR and scope the follow-up work appropriately :)
@EisenbergEffect
